Watch permissions

zpytela opened the pull request:
Introduce new permissions for the classes file and filesystem: watch watch_mount watch_reads watch_sb watch_with_perm
```
0c97d3925 Allow policykit various watch permissions
d620b3848 Allow dnsmasq watch systemd-resolved pid directories
49d0a1a3b Allow session_bus_type watch directories in /etc/dbus-1 and /usr
3d73abf0c Add interfaces to watch various files and directories
cd255e10f Allow nscd watch files and directories in /etc
0c384b2d9 Allow init_t watch /var/run/systemd/ask-password
0ce9ba052 Allow firewalld watch directories in /etc/firewalld
75080145b Allow login_userdomain watch various files/symlinks/dirs
c5db9d6e1 Add various systemd-related watches
beba762f1 Allow systemd-dbusd watch directories in /etc/dbus-1 and /usr
c9f17432e Allow sssd watch various files/symlinks/dirs
84d08c999 Allow syslogd watch its private runtime dirs and generic log dirs
35eb390f0 Allow init_t watch init_var_run_t directories
59182f053 Allow getty_t watch getty_var_run_t files
40bdc69de Disable automatic watch permissions for init_t
709de9cb9 Allow init_t to watch mount directories
7dc5adde6 Allow init_t to watch localization symlinks
843367009 Allow init_t to watch cgroup files
4a4761a66 Add basic watch patterns
c0135bbb2 Disable automatic watch permissions for unconfined domains
72ff7700a Set up scaffolding for the new watch permissions
32d6ecda5 Add new watch* permissions
```

WOnder93 committed:
Add new watch* permissions

Introduced in kernel commit ac5656d8a4cd ("fanotify, inotify, dnotify,
security: add security hook for fs notifications") [1], which is
currently staged for inclusion in the 5.4 kernel release.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/pcmoore/selinux.git/commit/?h=next&id=ac5656d8a4cdd93cd2c74355ed12e5617817e0e7

WOnder93 committed:
Add basic watch patterns

WOnder93 committed:
Allow init_t to watch cgroup files

WOnder93 committed:
Allow init_t to watch localization symlinks

It needs to watch (at least) /etc/localtime for changes.

WOnder93 committed:
Allow init_t to watch mount directories

It needs to watch the /run/mount directory.

zpytela commented:
This is a draft PR. It is not expected to be used for other then testing purpose to gather additional denials.
Further notes:
- temporarily, there are generic auditallow rules for the watch permissions to keep the system working
- additional permissions are known missing, they will be added later to make this PR more comprehensible (abrt*_t colord_t blkmapd_t rpcd_t gssd_t avahi_t vdagent_t)
There are copr-builds available on regular basis, the latest one confirmed working is selinux-policy-3.14.7-14.205.watch.fc34, with auditallow rules. Use the following commands:
```
dnf copr enable zpytela/selinux-policy
dnf update selinux policy
```

zpytela commented:
Just added
```
739e0a897 Allow init_t watch_reads init_var_run_t directories
```
I am aware the watch_reads perm may be later added as an additional one to an existing interface. This can happen to other commits as well, and anyway the complete patchset will be a subject of reviewing and rearranging once majority of permissions is gathered

There is now `selinux-policy-3.14.7-14.206.watch.fc34` copr-build available.

WOnder93 commented:
How about we create a `watch-permissions` branch directly in this repo so that anyone can send PRs against that branch? Then we can also set up a COPR repo to rebuild the latest version automatically.

(Any in-progress PRs will need careful rebasing whenever the branch would be rebased or some commits squashed, but that's not a big deal IMHO.)

zpytela commented:
The latest copr-build is selinux-policy-3.14.7-16.301.watch.fc34. System boots and runs without a denial, but not all services were examined.
Permissions for systemd are enumeated although allowing watch for file_type (or e. g. non_security_file_type) is still under consideration.

zpytela committed:
Add additional watch object permissions sets and patterns

The following file and directory object permissions sets and
dir patterns were added:

watch_mount_dir_perms
watch_reads_dir_perms
watch_with_perm_dir_perms
watch_reads_file_perms
watch_reads_dirs_pattern

zpytela committed:
Add basic watch interfaces to the base module

The following interfaces were added:

auth_watch_passwd
files_watch_root_dirs
files_watch_etc_dirs
files_watch_etc_files
files_watch_home
files_watch_usr_dirs
files_watch_var_run_dirs
files_watch_all_pid
fs_watch_rpc_dirs
fs_watch_mount_tmpfs_dirs
fs_watch_with_perm_tmpfs_dirs
init_watch_utmp
init_watch_dir
libs_watch_lib_dirs
mount_watch_reads_pid_dirs
mount_watch_pid_files
mount_watch_reads_pid_files
sysnet_watch_config

zpytela committed:
Add basic watch interfaces for systemd

The following interfaces were added:

systemd_resolved_watch_pid_dirs
systemd_login_watch_pid_dirs
systemd_login_watch_session_dirs
systemd_passwd_watch_pid_dirs
systemd_machined_watch_pid_dirs
systemd_networkd_watch_pid_dirs

zpytela committed:
Add watch permissions for user domain types

These watch permissions were added for login_userdomain using
the following interfaces:
- fs_watch_cgroup_files
- miscfiles_watch_localization_symlinks
- mount_watch_pid_dirs
- mount_watch_pid_files
- mount_watch_reads_pid_files

Interface userdom_watch_tmp_files() was added.

zpytela committed:
Add watch permissions for syslogd

Allow syslogd watch syslogd runtime dir and generic log dirs.
Add the logging_watch_generic_log_dirs() interface.

zpytela committed:
Allow getty watch its private runtime files

zpytela committed:
Add watch permissions for systemd services

Allow systemd-logind watch utmp files.
Allow systemd-resolved watch systemd-networkd runtime directories.
Allow systemd-timedated watch systemd-networkd runtime directories.

zpytela committed:
Add watch permissions for firewalld and NetworkManager

zpytela committed:
Add watch permissions for nscd and sssd

zpytela committed:
Add watch permissions for contrib packages

Watch permissions were added for the following domains and attributes:
- accountsd_t
- blkmapd_t
- cachefilesd_t
- crond_t
- gpg_agent_t
- gssd_t
- policykit_t
- policykit_auth_t
- rpcd_t
- system_dbusd_t
- session_bus_type
- thumb_t

zpytela commented:
I have just grouped and reordered existing commits for better clarity. I'd appreciate a review now.

zpytela committed:
Add watch permissions for init

Allow init watch directories with etc_t, home_root_t, and root_t types.
Allow init watch files and directories with the pidfile attribute.
Allow init watch_reads mount PID directories.

zpytela committed:
Allow restorecond watch /etc dirs

zpytela committed:
Allow xdm watch /usr directories

zpytela commented:
The latest copr-build is selinux-policy-3.14.7-16.305.watch.fc34. System boots without an AVC denial.

System also boots with the mls policy without watch-related AVC denial. Some additional denials will likely pop up for domains like sysadm_dbusd_t.

Koncpa commented on policy/modules/system/systemd.if:
```diff
@@ -1047,6 +1104,24 @@ interface(`systemd_relabelto_fifo_file_passwd_run',`
 
+########################################
+## <summary>
+##	Watch
```
Maybe write more about this interface in summary

Koncpa commented on policy/modules/system/miscfiles.if:
```diff
@@ -536,6 +536,31 @@ interface(`miscfiles_read_localization',`
 
+########################################
+## <summary>
+##	Allow process to watch localization symlnks.
```
*symlinks

zpytela commented:
Fixed, thank you.

zpytela commented:
I am going to merge this PR, all tests results are promising.
